compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala

```diff
@@ -602,16 +613,67 @@ private class ExtractAPICollector(using Context) extends ThunkHolder {
     annots.toList
   }
 
+  /** Produce a hash for a tree that is as stable as possible:
+   *  it should stay the same across compiler runs, compiler instances,
+   *  JVMs, etc.
+   */
+  def treeHash(tree: Tree): Int =
+    import scala.util.hashing.MurmurHash3
+
+    def positionedHash(p: ast.Positioned, initHash: Int): Int =
+      p match
+        case p: WithLazyField[?] =>
+          p.forceIfLazy
+        case _ =>
+      // FIXME: If `p` is a tree we should probably take its type into account
+      // when hashing it, but producing a stable hash for a type is not trivial
+      // since the same type might have multiple representations, for method
+      // signatures this is already handled by `computeType` and the machinery
+      // in Zinc that generates hashes from that, if we can reliably produce
+      // stable hashes for types ourselves then we could bypass all that and
+      // send Zinc hashes directly.
+      val h = MurmurHash3.mix(initHash, p.productPrefix.hashCode)
+      iteratorHash(p.productIterator, h)
+    end positionedHash
+
+    def iteratorHash(it: Iterator[Any], initHash: Int): Int =
+      var h = initHash
+      while it.hasNext do
+        it.next() match
+          case p: Positioned =>
+            h = positionedHash(p, h)
+          case xs: List[?] =>
+            h = iteratorHash(xs.iterator, h)
+          case c: core.Constants.Constant =>
+            h = MurmurHash3.mix(h, c.tag)
+            h = MurmurHash3.mix(h, c.value.##) // We can't use `value.hashCode` since value might be null
```

Comment:
is this ok to deal with types for `ClazzTag`? - i ignore symbols as `EnumTag` was removed upstream

Reply:
Hmm, good point, I guess we can call `apiType` to get a value with a stable hash which should be good enough